Save more file information from initial call...

rocky opened the pull request:
In particular, resolve the initial source directory and store that as _Dbg_init_dir. Resolve the name of main bash program and add that to canonicalized filenames.

@jansorg I think this fixes the second problem you mentioned in #50 Please double check though.

rocky committed:
Save more file information from initial call...

In particular, resolve the initial source directory and store that as
_Dbg_init_dir. Resolve the name of main bash program and add that to
canonicalized filenames.

jansorg commented:
@rocky Unfortunately, this did not fix the problem.
I've attached my debug session, also showing file location and content.

jansorg commented:
@rocky Thank you, this is fixing the problem!
Unfortunately, it seems to add a regression that break with a relative path doesn't work anymore, e.g. `break ./sources/lib.bash:2`.
`break`  with an absolute path is working well.

I've attached a screenshot of my session.

![image](https://github.com/user-attachments/assets/61858e43-125e-4f17-a356-7d1d9f88d9ce)

jansorg commented on lib/hook.sh:
```diff
@@ -119,6 +119,13 @@ _Dbg_debug_trap_handler() {
     _Dbg_save_args "$@"
 
+    typeset _Dbg_full_filename
+    _Dbg_full_filename=$(_Dbg_is_file "${BASH_SOURCE[1]}")
```
nit: quoting
```suggestion
    _Dbg_full_filename="$(_Dbg_is_file "${BASH_SOURCE[1]}")"
```

jansorg commented on bashdb.in:
```diff
@@ -102,6 +102,15 @@ if [[ -n $_Dbg_script_file ]] ; then
 
+# Resolve and save mapping for main script, and resolve
+# the starting directory.
+_Dbg_full_filename=$(_Dbg_is_file "$_Dbg_script_file")
```
nit: quoting
```suggestion
_Dbg_full_filename="$(_Dbg_is_file "$_Dbg_script_file")"
```

jansorg commented on bashdb.in:
```diff
@@ -102,6 +102,15 @@ if [[ -n $_Dbg_script_file ]] ; then
+# Note: expand_filename is expanding a *directory* here, not a filename.
+# This might cause a problem in the future if _Dbg_expand_filename becomes
+# more specific about this aspect.
+_Dbg_init_cwd=$(_Dbg_expand_filename "${_Dbg_script_file%/*}")
```
nit: quoting
```suggestion
_Dbg_init_cwd="$(_Dbg_expand_filename "${_Dbg_script_file%/*}")"
```

rocky committed:
Update bashdb.in

Co-authored-by: Joachim Ansorg <github@ja-dev.eu>

rocky committed:
Update lib/hook.sh

Co-authored-by: Joachim Ansorg <github@ja-dev.eu>

rocky committed:
Update bashdb.in

Co-authored-by: Joachim Ansorg <github@ja-dev.eu>

rocky commented:
> @rocky Thank you, this is fixing the problem! Unfortunately, it seems to add a regression that break with a relative path doesn't work anymore, e.g. `break ./sources/lib.bash:2`. `break` with an absolute path is working well.
> 
> I've attached a screenshot of my session.
> 
> ![image](https://private-user-images.githubusercontent.com/11473063/396455338-61858e43-125e-4f17-a356-7d1d9f88d9ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0MzI5NjEsIm5iZiI6MTczNDQzMjY2MSwicGF0aCI6Ii8xMTQ3MzA2My8zOTY0NTUzMzgtNjE4NThlNDMtMTI1ZS00ZjE3LWEzNTYtN2QxZDlmODhkOWNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDEwNTEwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0YWZiZTAwYmU2MGU2NWIyMTlkNmIxZDcxYTQxNmE2OTIwOWNjYjBkZGQ3NzIwM2Y3ZTZhNTFjNmQ3NzQ1MWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vikpvgoPvXvc6k5B3Sqce0OyCRUHuvTQruNxtLiZ5oE)

Thanks for checking and the changes. I will try to look at and address this problem over the weekend.

rocky commented:
> > @rocky Thank you, this is fixing the problem! Unfortunately, it seems to add a regression that break with a relative path doesn't work anymore, e.g. `break ./sources/lib.bash:2`. `break` with an absolute path is working well.
> > I've attached a screenshot of my session.
> > ![image](https://private-user-images.githubusercontent.com/11473063/396455338-61858e43-125e-4f17-a356-7d1d9f88d9ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0MzI5NjEsIm5iZiI6MTczNDQzMjY2MSwicGF0aCI6Ii8xMTQ3MzA2My8zOTY0NTUzMzgtNjE4NThlNDMtMTI1ZS00ZjE3LWEzNTYtN2QxZDlmODhkOWNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDEwNTEwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0YWZiZTAwYmU2MGU2NWIyMTlkNmIxZDcxYTQxNmE2OTIwOWNjYjBkZGQ3NzIwM2Y3ZTZhNTFjNmQ3NzQ1MWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vikpvgoPvXvc6k5B3Sqce0OyCRUHuvTQruNxtLiZ5oE)
> 
> Thanks for checking and the changes. I will try to look at and address this problem over the weekend.

@jansorg When I went to look at this over the weekend the image was missing. Would you upload it again or give a sequence of instructions? 

<strike>When I debug I see that in relative path resolution favor the initial current working directory over the current working directory, and that can sometimes be wrong. But it is hard to know if this is exactly the problem. </strike> Nevermind. I see that this is only an image problem in my reply, not what you posted originally.

The key issue is that the "break" command is not finding the file when using a relative path. As you report, using an absolute path works fine.

rocky committed:
Add more places we resolve/expand files

In some cases though automatic file loading no longer happens.
In particular, on "list" and "break" commands.

rocky commented:
@jansorg with 896a12f the example you give now works. However, this also disables the automatic file loading of files with relative file names not previously encountered. 

Instead, you must specify an explicit "load" here. The automatic loads of relative file names happened in the "list" and "break" commands. With this change, you will get an "internal error" message instead. 

Addressing that will take more work and effort. I am not sure though it is that important. What do you think?

jansorg commented:
I'll be able to test your changes end of this week. Thank you for looking into this!

No automatic loading of relative paths seems okay to me, unless I misunderstood what how it works.
As far as I understand, instead of a "break ./a/b.sh" an additional "load ./a/b.sh" is needed first. I didn't even know that this was a thing and my IDE integration is always executing a "load" command first in such a case.

jansorg commented on lib/fns.sh:
```diff
@@ -197,6 +197,9 @@ function _Dbg_linespec_setup {
   line_number=${word[0]}
   full_filename=$(_Dbg_is_file "$filename")
+  if [[ -z "$full_filename" ]] ; then
+      full_filename=$(_Dbg_resolve_expand_filename "$filename")
```
Is quoting needed here?

rocky replied:
I tested with quoting and it doesn't seem to hurt. Personally, I find the quotes within quotes confusing, and more generally the multiple substitution passes weird, but this is bash after all. 

So I've now made the change.

rocky commented:
> I'll be able to test your changes end of this week. Thank you for looking into this!
> 
> No automatic loading of relative paths seems okay to me, unless I misunderstood what how it works. As far as I understand, instead of a "break ./a/b.sh" an additional "load ./a/b.sh" is needed first. I didn't even know that this was a thing and my IDE integration always executes a "load" command first in such a case.

In stepping, the debugger may encounter relative paths in BASH_SOURCE. It turns these into absolute paths and saves the mapping as soon as possible,  to avoid the problem of the current working directory changing inside the debugged program. So in that respect, the auto conversion always happens, and has to happen.

In addition, a user can issue debugger commands like "list" or "break"  fabricating any name the user wants.  It is here that I guess we hitting places where we are not converting the filename to its full name or storing canonical information about the file.

rocky committed:
One more quoted statement.

jansorg commented:
I've tested your branch and as far as I can tell, it's working great! Thank you for fixing this problem!
